plugins/SlicerT/WaveForm.cpp

```diff
@@ -0,0 +1,333 @@
+/*
+ * WaveForm.cpp - slice editor for SlicerT
+ *
+ * Copyright (c) 2006-2008 Daniel Kauss Serna <daniel.kauss.serna@gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "WaveForm.h"
+
+#include "SlicerT.h"
+#include "embed.h"
+
+namespace lmms {
+
+namespace gui {
+
+WaveForm::WaveForm(int w, int h, SlicerT* instrument, QWidget* parent)
+	: QWidget(parent)
+	,
+	// calculate sizes
+	m_width(w)
+	, m_height(h)
+	, m_seekerWidth(w - m_seekerHorMargin * 2)
+	, m_editorHeight(h - m_seekerHeight - m_middleMargin)
+	, m_editorWidth(w)
+	,
+
+	// create pixmaps
+	m_sliceArrow(PLUGIN_NAME::getIconPixmap("slide_indicator_arrow"))
+	, m_seeker(QPixmap(m_seekerWidth, m_seekerHeight))
+	, m_seekerWaveform(QPixmap(m_seekerWidth, m_seekerHeight))
+	, m_sliceEditor(QPixmap(w, m_editorHeight))
+	,
+
+	// references to instrument vars
+	m_currentSample(instrument->m_originalSample)
+	, m_slicerTParent(instrument)
+	, m_slicePoints(instrument->m_slicePoints)
+
+{
+	// window config
+	setFixedSize(m_width, m_height);
+	setMouseTracking(true);
+
+	// draw backgrounds
+	m_sliceEditor.fill(m_waveformBgColor);
+	m_seekerWaveform.fill(m_waveformBgColor);
+
+	// connect to playback
+	connect(m_slicerTParent, SIGNAL(isPlaying(float, float, float)), this, SLOT(isPlaying(float, float, float)));
+	connect(m_slicerTParent, SIGNAL(dataChanged()), this, SLOT(updateUI()));
+
+	updateUI();
+}
+
+void WaveForm::drawSeekerWaveform()
+{
+	m_seekerWaveform.fill(m_waveformBgColor);
+	if (m_currentSample.frames() < 2048) { return; }
```

Comment:
Why are we preventing certain samples from executing most of the functions if they don't have at least 2048 frames? I would think this is a workaround for some nasty bug elsewhere.

Reply:
I used this generally to detect if a sample is even loaded. If no sample is loaded, the functions don't execute. Why 2048? because a sample that's shorter would probably break some code.  I haven't really tested it which such sort samples, but who wants to slice a 0.04 second long sample anyway.

Reply:
> I used this generally to detect if a sample is even loaded. If no sample is loaded, the functions don't execute. Why 2048? because a sample that's shorter would probably break some code. I haven't really tested it which such sort samples, but who wants to slice a 0.04 second long sample anyway.

This seems really hacky. A 0.04 second long sample is still a valid sample, and there are better ways of checking it. Cant you just use a ``bool m_fileLoaded`` or something? When resetting/unloading the sample, you could just set it to ``false``, and to ``true`` vice versa.

Reply:
Is it really that hacky to set a minimum sample size? I can reduce it a bit, but I think checking for the sample like this is reasonable. AFP and the LFO also use this as a check, although with a min size of 1. And again, such a short sample is extremely rare, and would not fit within the context of a slicer, because what is there even to slice? If this is really such a big problem, I can change it, but the check always ensures correct functionality, with pretty much no drawbacks.

Reply:
> Is it really that hacky to set a minimum sample size? I can reduce it a bit, but I think checking for the sample like this is reasonable. AFP and the LFO also use this as a check, although with a min size of 1. And again, such a short sample is extremely rare, and would not fit within the context of a slicer, because what is there even to slice? If this is really such a big problem, I can change it, but the check always ensures correct functionality, with pretty much no drawbacks.

![image](https://github.com/LMMS/lmms/assets/20197645/678c74b9-a7e3-4982-a5f2-93d974c9f306)

Take Ninja for instance. Even if there's essentially no sound playing per slice, it stills allows me to load the 0.04 second long sample in. In practice, I do agree with you that a 0.04 second sample is hard to come by, and no one would really try to slice that. But, in terms of the code not blindly filtering a collection of samples out because they don't meet the duration (or even worrying about the duration entirely), I personally think it makes the code more robust.

It just feels a bit unnecessary to filter out such samples. What harm is there in loading a 0.04 second sample? Does some part of the code need to meet a certain number of frames for it to work properly?

Reply:
> It just feels a bit unnecessary to filter out such samples. What harm is there in loading a 0.04 second sample? Does some part of the code need to meet a certain number of frames for it to work properly?

Well yes, the sample has to be bigger than the FFTs window sizes, which are 512 for the PV and 1024 for the beat detection. Of course this could be removed by padding the input and whatever, but its just simpler to not allow samples less than 2048 frames. I will check with small samples and see if they work, but I really don't want to rewrite a lot of code just so you can load a 512 frame sample or whatever. I think setting reasonable requirements for proper behaviour is not at all wrong.

Reply:
> Well yes, the sample has to be bigger than the FFTs window sizes, which are 512 for the PV and 1024 for the beat detection. Of course this could be removed by padding the input and whatever, but its just simpler to not allow samples less than 2048 frames. I will check with small samples and see if they work, but I really don't want to rewrite a lot of code just so you can load a 512 frame sample or whatever. I think setting reasonable requirements for proper behaviour is not at all wrong.

Ninjas2 uses aubio and is not met with such a restriction. I'm guessing the way things are parametrized between your slicer and Ninjas are just different? However, in that case, then I suppose that's a valid reason to prevent short samples from loading.

In fact, with the sample shown previously, the onsets option does not do anything, but the raw option does. I think maybe you could enable loading the samples in, but just disable beat detection, but if your slicer is structured in such a way that there is no support to evenly slice samples (like with the raw option in Ninjas) and only has support for onsets, then I guess we can leave the duration limit if it would make things simpler.